### PR TITLE
Hydra migration 2

### PR DESCRIPTION
**Pull request by vincentpierre**
## Motivation and Context

Second batch of changes for hydra migration
Currently breaking all tests
Changes the way Sensors, Camera Sensors, Action and Measures are passed in the config. 
Needs changes to all existing configs
` test/test_baselines_hydra.py` is passing

## Types of changes

- Docs change / refactoring / dependency upgrade
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py [hunk 1]
```diff
@@ -150,14 +150,20 @@ class Eq2CubeConfig(ObsTransformConfig):
     )
 
 
+@dataclass
+class TmpObsTransformConfig(HabitatBaselinesBaseConfig):
+    # TODO : Make this a dict of str:ObsTransformConfig rather than use an enabled_transforms list
+    enabled_transforms: List[str] = field(default_factory=lambda: [])
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/config/default_structured_configs.py:156]: Currently a hack. Needs to change later. This is to get around the `enabled_transforms` which will not exist in the future.

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py [hunk 2]
```diff
@@ -200,6 +206,11 @@ class VERConfig(HabitatBaselinesBaseConfig):
     overlap_rollouts_and_learn: bool = False
 
 
+@dataclass
+class TmpAuxLossConfig(HabitatBaselinesBaseConfig):
+    enabled: List[str] = field(default_factory=lambda: [])
```
Comment (vincentpierre) [inline habitat-baselines/habitat_baselines/config/default_structured_configs.py:211]: Same as TmpObsTransformConfig. This will eventually go away

File: habitat-lab/habitat/config/default_structured_configs.py [hunk 3]
```diff
@@ -699,28 +699,21 @@ class TaskConfig(HabitatBaseConfig):
...
-            "turn_left",
-            "turn_right",
-        ]
-    )
     # REARRANGE task
     count_obj_collisions: bool = True
-    settle_steps = 5
-    constraint_violation_ends_episode = True
-    constraint_violation_drops_object = False
+    settle_steps: int = 5
```
Comment (vincentpierre) [inline habitat-lab/habitat/config/default_structured_configs.py:708]: Needed typing for it to be recognized by Hydra

File: habitat-lab/habitat/config/default_structured_configs.py [hunk 4]
```diff
@@ -879,7 +872,7 @@ class ThirdDepthSensorConfig(HabitatSimDepthSensorConfig):
 class AgentConfig(HabitatBaseConfig):
     height: float = 1.5
     radius: float = 0.1
-    sensors: List[str] = field(default_factory=lambda: ["rgb_sensor"])
+    camera_sensors: Dict[str, Dict[str, Any]] = field(default_factory=dict)
```
Comment (vincentpierre) [inline habitat-lab/habitat/config/default_structured_configs.py:875]: Temporary. I wanted to do `Dict[str, SensorConfig]` but this does not work as some keys are present in some SensorConfig implementations but not in the base SensorConfig. Will need to figure out a workaround.

File: habitat-lab/habitat/core/embodied_task.py [hunk 5]
```diff
@@ -234,38 +235,31 @@ def __init__(
...
-        self, entity_names, register_func, entities_config=None
-    ) -> OrderedDict:
-        if entities_config is None:
-            entities_config = self._config
+    def _init_entities(self, entities_configs, register_func) -> OrderedDict:
 
         entities = OrderedDict()
-        for entity_name in entity_names:
-            entity_cfg = getattr(entities_config, entity_name)
+        for entity_name, entity_cfg in entities_configs.items():
```
Comment (vincentpierre) [inline habitat-lab/habitat/core/embodied_task.py:261]: Big change that breaks all existing yaml (except for the pick.yaml that was converted). Instead of a dict + list of enabled, now we only have a dict.

**Comment by dhruvbatra** [comment]
> Working on it. Harder than I thought. Regretting life choices](https://github.com/facebookresearch/habitat-lab/pull/983/commits/3356f31bc298a60d62ab0ec6d80876a69bcbb411)

👀 😂

File: habitat-lab/habitat/config/habitat/tasks/imagenav.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
+# @package habitat
+
+  defaults:
+    - /habitat/habitat_config_base@
```
Comment (rpartsey) [inline habitat-lab/habitat/config/habitat/tasks/imagenav.yaml:4]: Remove trailing `@`

File: habitat-lab/habitat/tasks/rearrange/multi_task/task_creator_utils.py [hunk 7]
```diff
@@ -43,21 +45,25 @@ def create_task_object(
...
+        task_config = habitat.get_config(
+            osp.join(TASK_CONFIGS_DIR, task_config_path + ".yaml"),
+            pass_args,
+        )
+
+        with habitat.config.read_write(config):
+            config = OmegaConf.merge(config, task_config.habitat.task)
+            # config.merge_from_other_cfg(task_config.habitat.task)
+            # Putting back the values from TASK_IGNORE_KEYS :
             for k in TASK_IGNORE_KEYS:
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/multi_task/task_creator_utils.py:62]: @ASzot I changed the way the `TASK_IGNORE_KEYS` are handled. Instead of removing them from the overriding config, I re-add them from the original base configuration. Let me know if this is not the behavior that was intended.

File: .circleci/config.yml [hunk 8]
```diff
@@ -149,7 +149,7 @@ jobs:
                 . activate habitat
                 pip install -U pip
                 conda install -q -y -c conda-forge ninja ccache numpy pytest pytest-mock pytest-cov psutil
-                pip install pytest-sugar mock cython pygame
```
Comment (Skylion007) [inline .circleci/config.yml:152]: pytest-sugar should be fixed now.

File: habitat-baselines/habitat_baselines/config/default_structured_configs.py [hunk 9]
```diff
@@ -30,14 +32,15 @@ class WBConfig(HabitatBaselinesBaseConfig):
...
-    # is useful if some code changes necessitate a new config but the weights
-    # are still valid.
     use_ckpt_config: bool = True
     should_load_ckpt: bool = True
     # The number of time to run each episode through evaluation.
     # Only works when evaluating on all episodes.
     evals_per_ep: int = 1
+    video_option: List[str] = field(
+        # available options are "disk" and "tensorboard"
+        default_factory=lambda: []
```
Comment (Skylion007) [inline habitat-baselines/habitat_baselines/config/default_structured_configs.py:42]:
```suggestion
        default_factory=list
````

Comment (vincentpierre) [reply habitat-baselines/habitat_baselines/config/default_structured_configs.py:42]: Thanks !

File: habitat-lab/habitat/config/default_structured_configs.py [hunk 10]
```diff
@@ -782,12 +780,14 @@ class HabitatSimRGBSensorConfig(SimulatorCameraSensorConfig):
 
 
 @dataclass
-class HabitatSimDepthSensorConfig(SimulatorDepthSensorConfig):
+class HabitatSimDepthSensorConfig(
```
Comment (Skylion007) [inline habitat-lab/habitat/config/default_structured_configs.py:783]: It's a broader refactor question if we want different dataclasses for each sensor type or if the config should be more general. Regardless, that is orthogonal to this PR just to keep in mind.

Comment (vincentpierre) [reply habitat-lab/habitat/config/default_structured_configs.py:783]: Yes, we should have this discussion. Having a dataclass for each sensor allows to have type checking for each sensor, which actually helped us catch a few issues.

**Comment by Skylion007** [comment]
BTW, any reason we are using dataclass all overall the place instead of attr.s (which is already a habitat-lab requirement). Hydro supports both: See https://www.attrs.org/en/stable/why.html#data-classes for the subtle differences.

File: habitat-lab/habitat/config/default_structured_configs.py (hunk 4 above)
Comment (Skylion007) [reply habitat-lab/habitat/config/default_structured_configs.py]: This is is a typing issue /architecture issue. @vincentpierre the solution here is to specify a Union across all the ivalid implementations of Sensor Ccnfigs or to redo the hierarchy to add abstract methods the support it as some intermediate ancestor / mixin

**Comment by vincentpierre** [comment]
> BTW, any reason we are using dataclass all overall the place instead of attr.s (which is already a habitat-lab requirement). Hydro supports both: See https://www.attrs.org/en/stable/why.html#data-classes for the subtle differences.
> Biggest reason is that attr.s could pass slots=True which would make attribute lookup on the configs a bit faster.

Will investigate this. I think this is fine merging now and look into this in the next PR. The goal of this Pr is to get hydra working, optimizations (and there will be a bunch) will come later.
Thanks for helping up with this!

File: habitat-lab/habitat/config/default_structured_configs.py (hunk 4 above)
Comment (rpartsey) [reply habitat-lab/habitat/config/default_structured_configs.py]: @Skylion007 This is no longer the Issue. The highlighted code is outdated. Everything works fine with `Dict[str, SensorConfig]`.

**Comment by rpartsey** [comment]
> BTW, any reason we are using dataclass all overall the place instead of attr.s (which is already a habitat-lab requirement). Hydro supports both: See https://www.attrs.org/en/stable/why.html#data-classes for the subtle differences.
> 
> Biggest reason is that attr.s could pass slots=True which would make attribute lookup on the configs a bit faster.

Thank you for the comment @Skylion007!
According to the [OmegaConf](https://omegaconf.readthedocs.io/en/2.2_branch/structured_config.html) docs, they don't differentiate between `@dataclass` and `attr.s`. They say that [dataclasses](https://docs.python.org/3.7/library/dataclasses.html) are standard as of Python 3.7 and you don't need to install any additional packages to support structured configs. But, yes if `attr.s` can speedup execution its definitely worth considering.

I'm not sure if slots will attribute lookup faster as structured configs are used as a validation schemas to create output config. After that, all the attribute access is performed on the output config.

@vincentpierre can we ask Hydra team what is the best choice for structured config definition?

File: habitat-lab/habitat/tasks/rearrange/multi_task/task_creator_utils.py [hunk 11]
```diff
@@ -43,21 +45,25 @@ def create_task_object(
...
-            pass_args: List[Any] = []
-            for k, v in task_config_args.items():
-                pass_args.extend((k, v))
-            task_config = habitat.get_config(
-                osp.join(TASK_CONFIGS_DIR, task_config_path + ".yaml"),
-                pass_args,
-            )
+
+    if task_config_path is not None:
+        pass_args: List[Any] = []
```
Comment (Skylion007) [inline habitat-lab/habitat/tasks/rearrange/multi_task/task_creator_utils.py:50]:
```suggestion
        pass_args: List[str] = []
```

File: habitat-lab/habitat/tasks/rearrange/multi_task/task_creator_utils.py [hunk 12]
```diff
@@ -43,21 +45,25 @@ def create_task_object(
...
+        for k, v in task_config_args.items():
+            pass_args.append(f"{k}={v}")
```
Comment (Skylion007) [inline habitat-lab/habitat/tasks/rearrange/multi_task/task_creator_utils.py:52]:
```suggestion
        pass_args: List[str] = [f"{k}={v}" for k, v in task_config_args.items()]
```
faster, more efficient, also I think this makes the explicit specifying of the List type unnecessary.

File: habitat-lab/habitat/tasks/rearrange/rearrange_sim.py [hunk 13]
```diff
@@ -48,18 +48,16 @@ class RearrangeSim(HabitatSim):
...
-                    for orig_sensor_name in config.agent_0.sensors:
-                        full_name = f"{agent}_{orig_sensor_name}"
-                        orig_sensor_id = config[orig_sensor_name].uuid
-                        new_sensor_cfg = config[orig_sensor_name].clone()
-                        new_sensor_cfg.uuid = f"{agent}_{orig_sensor_id}"
-                        config[full_name] = new_sensor_cfg
-                        all_new_sensor_names.append(full_name)
-
-                config.agent_0.sensors = all_new_sensor_names
+                    sensor_keys = list(agent_cfg.sim_sensors.keys())
```
Comment (Skylion007) [inline habitat-lab/habitat/tasks/rearrange/rearrange_sim.py:53]:
```suggestion
                    sensor_keys = agent_cfg.sim_sensors.keys()
```
No reason this needs to be a list. It is only iterated over once.

**Review by Skylion007 — COMMENTED** [review]
Some minor nits, but overall looks good.

File: habitat-lab/habitat/core/utils.py [hunk 14]
```diff
@@ -114,25 +115,30 @@ def center_crop(obs, new_shape):
...
     def default(self, obj):
-        # JSON doesn't support numpy ndarray and quaternion
         if isinstance(obj, np.ndarray):
             return obj.tolist()
         if isinstance(obj, quaternion.quaternion):
             return quaternion_to_list(obj)
+        if isinstance(obj, DictConfig):
+            return dict(obj)
+        if isinstance(obj, ListConfig):
+            return list(obj)
```
Comment (Skylion007) [inline habitat-lab/habitat/core/utils.py:127]:
```suggestion
        if OmegaConf.is_config(obj)
            return OmegaConfg.to_container(obj)
```
This should be much more efficient.

File: habitat-lab/habitat/core/utils.py [hunk 15]
```diff
@@ -12,6 +12,7 @@
 import attr
 import numpy as np
 import quaternion  # noqa: F401
+from omegaconf import DictConfig, ListConfig
```
Comment (Skylion007) [inline habitat-lab/habitat/core/utils.py:15]:
```suggestion
from omegaconf import OmegaConf
```

**Comment by Skylion007** [comment]
> > BTW, any reason we are using dataclass all overall the place instead of attr.s (which is already a habitat-lab requirement). Hydro supports both: See https://www.attrs.org/en/stable/why.html#data-classes for the subtle differences.
> > Biggest reason is that attr.s could pass slots=True which would make attribute lookup on the configs a bit faster.
> 
> Thank you for the comment @Skylion007! According to the [OmegaConf](https://omegaconf.readthedocs.io/en/2.2_branch/structured_config.html) docs, they don't differentiate between `@dataclass` and `attr.s`. They say that [dataclasses](https://docs.python.org/3.7/library/dataclasses.html) are standard as of Python 3.7 and you don't need to install any additional packages to support structured configs. But, yes if `attr.s` can speedup execution (and its already a dependency) its definitely worth considering.
> 
> I'm not sure if slots will make attribute lookup faster as structured configs are used as a validation schemas to create output config. After that, all the attribute access is performed on the output config.
> 
> @vincentpierre can we ask Hydra team what is the best choice for structured config definition?

Seems like we have some nice benefits if we use .to_object() to store the dataclass in various places where we need it. Doing that will convert the config object to a nice list / dict of structured configs converted into their backed dataclass

**Review by rpartsey — APPROVED** [review]

File: habitat-baselines/habitat_baselines/rl/ppo/policy.py [hunk 16]
```diff
@@ -90,23 +90,23 @@ def __init__(
                 policy_config.action_dist,
             )
         else:
-            ValueError(
+            raise ValueError(
```
Comment (Skylion007) [inline habitat-baselines/habitat_baselines/rl/ppo/policy.py:93]: Huh? We have flake8 rules that should have caught this...

File: habitat-baselines/habitat_baselines/rl/hrl/skills/nn_skill.py [hunk 17]
```diff
@@ -154,7 +154,7 @@ def from_config(
             OrderedDict(
                 [
                     (k, action_space[k])
-                    for k in policy_cfg.habitat.task.possible_actions
+                    for k in policy_cfg.habitat.task.actions.keys()
                 ]
```
Comment (Skylion007) [inline habitat-baselines/habitat_baselines/rl/hrl/skills/nn_skill.py:158]:
```suggestion
                (
                    (k, action_space[k])
                    for k in policy_cfg.habitat.task.actions.keys()
                )
``` 
Pretty sure this can be a generator comprehension

File: examples/register_new_sensors_and_measures.py [hunk 18]
```diff
@@ -74,32 +88,41 @@ def get_observation(
...
 
     with habitat.Env(config=config) as env:
         print(env.reset()["agent_position"])
         print(env.get_metrics()["episode_info"])
+        # After reset my_value should be set
+        assert env.get_metrics()["episode_info"]["my_value"] == my_value
         print(env.step("move_forward")["agent_position"])
         print(env.get_metrics()["episode_info"])
+        # my_value should only be present at reset, not after step
+        assert "my_value" not in env.get_metrics()["episode_info"]
```
Comment (mathfac) [inline examples/register_new_sensors_and_measures.py:125]: I live how this example became even shorter.

**Review by mathfac — APPROVED** [review]
Thank you for putting together such significant change and following on all the tests. The all changes look reasonable and sure we can make more changes and improvements after merging it.

File: examples/benchmark.py [hunk 19]
```diff
@@ -24,7 +24,7 @@ def main():
     parser.add_argument(
         "--task-config",
         type=str,
-        default="tasks/pointnav.yaml",
+        default="benchmark/navigation/pointnav/pointnav_habitat_test.yaml",
```
[the remark this answers is not in the document]
Comment (vincentpierre) [reply examples/benchmark.py:27]: test here means the dataset is a demo dataset. We could rename pointnav_habitat_test.yaml to pointnav_habitat_demo.yaml in the future

**Review by Skylion007 — APPROVED** [review]
Except for my minor generator nit: this looks good to me for now.
